### PR TITLE
Introduce a separate slot for stealing tasks into in CoroutineScheduler

### LINKED ISSUE #3416: CoroutineScheduler: do not re-add stolen task to the head of the WorkerQueue where the task can be stolen from again
We should hand it directly to the stealing worker without intermediate stores.

Apart from code being more straightforward, it also saves us at least 2 CASes on a fairly contended memory location

### PR DESCRIPTION
**Pull request by qwwdfsad**
It solves two problems:

* Stealing into exclusively owned local queue does no longer require and CAS'es or atomic operations where they were previously not needed. It should save a few cycles on the stealing code path
* The overall timing perturbations should be slightly better now: previously it was possible for the stolen task to be immediately got stolen again from the stealer thread because it was actually published to owner's queue, but its submission time was never updated

Fixes #3416

### CODE REVIEW

File: kotlinx-coroutines-core/jvm/src/scheduling/CoroutineScheduler.kt [hunk 1]
```diff
@@ -873,15 +872,10 @@ internal class CoroutineScheduler(
...
+        fun findTask(mayHaveLocalTasks: Boolean): Task? {
+            if (tryAcquireCpuPermit()) return findAnyTask(mayHaveLocalTasks)
             // If we can't acquire a CPU permit -- attempt to find blocking task
-            val task = if (scanLocalQueue) {
-                localQueue.poll() ?: globalBlockingQueue.removeFirstOrNull()
-            } else {
-                globalBlockingQueue.removeFirstOrNull()
-            }
-            return task ?: trySteal(blockingOnly = true)
+            return globalBlockingQueue.removeFirstOrNull() ?: trySteal(blockingOnly = true)
```
Comment (qwwdfsad) [inline kotlinx-coroutines-core/jvm/src/scheduling/CoroutineScheduler.kt:878]: This fix seems to be incorrect, I will ament the last commit

Comment (qwwdfsad) [reply kotlinx-coroutines-core/jvm/src/scheduling/CoroutineScheduler.kt:878]: For each blocking task scheduled, the scheduler unparks/creates a separate worker to steal that the current thread blocks indefinitely (i.e. "no progress from here now on in this thread"), so no liveness issues or bugs, but it still worth trying to poll thread's local queue for blocking tasks only

File: kotlinx-coroutines-core/jvm/src/scheduling/WorkQueue.kt [hunk 2]
```diff
@@ -100,41 +101,37 @@ internal class WorkQueue {
...
-     * Tries stealing from [victim] queue into this queue.
+     * Tries stealing from this queue into the [stolenTaskRef] argument.
      *
      * Returns [NOTHING_TO_STEAL] if queue has nothing to steal, [TASK_STOLEN] if at least task was stolen
      * or positive value of how many nanoseconds should pass until the head of this queue will be available to steal.
      */
-    fun tryStealFrom(victim: WorkQueue): Long {
-        assert { bufferSize == 0 }
-        val task  = victim.pollBuffer()
+    fun trySteal(stolenTaskRef: ObjectRef<Task?>): Long {
```
Comment (dkhalanskyjb) [inline kotlinx-coroutines-core/jvm/src/scheduling/WorkQueue.kt:111]: The names no longer reflect who's the receiver and who's doing the action. The semantics of "I steal (from myself) (into this bag that was given to me)" is not really intuitive!
* `tryGift`? Semantically clear, but "stealing" is a well-established term.
* `tryStealFromThisInto`? Wordy but much less confusing.
* Make it a separate static function, like `trySteal(victim, stolenTaskRef)`.
* ...

File: kotlinx-coroutines-core/jvm/src/scheduling/WorkQueue.kt [hunk 3]
```diff
@@ -100,41 +103,63 @@ internal class WorkQueue {
...
+
+        while (start != end) {
+            --end
+            val index = end and MASK
+            if (blockingTasksInBuffer.value == 0) break
+            val value = buffer[index]
+            if (value != null && value.isBlocking && buffer.compareAndSet(index, value, null)) {
+                blockingTasksInBuffer.decrementAndGet()
+                return value
+            }
```
Comment (dkhalanskyjb) [inline kotlinx-coroutines-core/jvm/src/scheduling/WorkQueue.kt:160]: How about moving this into a separate method, like `tryExtractBlockingTask(index: Int)`?

File: kotlinx-coroutines-core/jvm/src/scheduling/WorkQueue.kt (hunk 2 above)
Comment (qwwdfsad) [reply kotlinx-coroutines-core/jvm/src/scheduling/WorkQueue.kt:111]: Maybe `tryStealInto()`? `victim.tryStealInto(stolenTaskSlot)` seems quite straightforward

File: kotlinx-coroutines-core/jvm/src/scheduling/WorkQueue.kt (hunk 3 above)
Comment (qwwdfsad) [reply kotlinx-coroutines-core/jvm/src/scheduling/WorkQueue.kt]: Done

File: kotlinx-coroutines-core/jvm/src/scheduling/WorkQueue.kt (hunk 2 above)
Comment (dkhalanskyjb) [reply kotlinx-coroutines-core/jvm/src/scheduling/WorkQueue.kt:111]: "I steal (into this bag)" begs the question, from whom do you steal? The answer is, from yourself, which is just not obvious. When I see `X.y()`, I always read this as "`X`, I order you to `y`".

Comment (qwwdfsad) [reply kotlinx-coroutines-core/jvm/src/scheduling/WorkQueue.kt:111]: I agree with the general reasoning. That's why I added context (receiver name), as this is the method used in a very specific manner on call sites we control, including the receiver's name.

I'm reluctant to both adding a companion (so the static method has FQN name like `WorkQueue.trySteal(victim, into)`) and to giving up on `steal` terminology

File: kotlinx-coroutines-core/jvm/src/scheduling/WorkQueue.kt [hunk 4]
```diff
@@ -100,41 +103,70 @@ internal class WorkQueue {
...
             val value = buffer[index]
             if (value != null && value.isBlocking && buffer.compareAndSet(index, value, null)) {
-                victim.blockingTasksInBuffer.decrementAndGet()
-                add(value)
+                blockingTasksInBuffer.decrementAndGet()
+                stolenTaskRef.element = value
                 return TASK_STOLEN
             } else {
                 ++start
             }
```
Comment (dkhalanskyjb) [inline kotlinx-coroutines-core/jvm/src/scheduling/WorkQueue.kt:134]:
```kotlin
val task = tryExtractBlockingTask(start)
if (task != null) {
  stolenTaskRef.element = task
  return TASK_STOLEN
} else {
  ++start
}
```
or even
```kotlin
stolenTaskRef.element = tryExtractBlockingTask(start++) ?: continue
return TASK_STOLEN
```
?

File: kotlinx-coroutines-core/jvm/src/scheduling/WorkQueue.kt (hunk 2 above)
Comment (dkhalanskyjb) [reply kotlinx-coroutines-core/jvm/src/scheduling/WorkQueue.kt:111]: Eh , ok. I think this warrants a comment though, in the vein of
"Counterintuitively, the queue from which a task is stolen is [this], so we are stealing from ourselves here. The reason for the name is, *stealing* is an established concept in concurrency".

**Pull request merged**